Set the focus on Edit Box after edit action.

parasharrajat opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
1. Remove modal visibility status from the Context Menu. which is needed to make the edited comment as a prime actor for focus.
2. Delayed the edit action after the Context menu is completely hidden which let us handle the focus. It removes the need for the SetTimeout hack.
3. Set the focus on the Edit box after each edit action.
4. Move the focus back to the main Composer when the Edit box is closed.

### Fixed Issues
<!-- Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing -->
Fixes #3017
fixes #3309 

### Tests | QA Steps (Web|Desktop)
* Sign in to the e.chat staging web app
* Right-click the text and choose Edit Comment
* Textbox should automatically get focused.
* Click Edit comment from Hover menu. Edit box should automatically get focused.
* Send a message, then hit the `ArrowUp` key. Edit box of your latest message should automatically get focused.
* Hit the `Esc` key, and the main `ReportActionCompose` should regain focus.

### Tests | QA Steps(mWeb|Android|IOS)
* Sign in to the e.chat staging web app
* Hold press any text message sent by you and choose Edit Comment
* Textbox should automatically get focused.

### Tests | QA Steps (Web|Desktop) for #3309
* Sign in to the e.chat staging web app
* Right-click the text and choose Edit Comment
* Textbox should automatically get focused.
* Click Edit comment from Hover menu. Edit box should automatically get focused.
* Save changes , main chat composer should regain focus.
* cancel Edit, main chat composer should regain focus.
* Click edit comment for any message.
* Hit the `Esc` key, and the main chat composer should regain focus.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [ ] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->


https://user-images.githubusercontent.com/24370807/121236427-93d76480-c8b3-11eb-9f20-9b8787a71e93.mp4


#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

https://user-images.githubusercontent.com/24370807/121238132-64c1f280-c8b5-11eb-868a-bbb082f05584.mp4


#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->


https://user-images.githubusercontent.com/24370807/121236814-fa5c8280-c8b3-11eb-9c5b-4afd93dfcc4d.mp4

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

https://user-images.githubusercontent.com/24370807/121239234-9ab3a680-c8b6-11eb-9209-a5ee0eaaade9.mp4

parasharrajat committed:
set the focus on the Edit ox on edit action

parasharrajat committed:
pull latest changes

parasharrajat committed:
fix: callback needs to reset for each action

parasharrajat committed:
Merge branch 'main' of github.com:Expensify/Expensify.cash into edit-box

parasharrajat committed:
added a new composer focus strategy

parasharrajat committed:
fix: bug transItionEnd does not fire on IOS

parasharrajat commented:
Ready to roll. Thanks.

roryabraham commented:
@parasharrajat Looks like lint is failing

parasharrajat commented:
@roryabraham this is due to https://github.com/Expensify/Expensify.cash/pull/3401. And the fix is being made here https://github.com/Expensify/Expensify.cash/pull/3481. So I suggest we merge it quickly.

roryabraham commented:
Lint error resolved, please merge `main` into this branch

roryabraham commented:
Also @parasharrajat will this also fix [this deploy blocker](https://github.com/Expensify/Expensify.cash/pull/3434)?

roryabraham commented:
> Also @parasharrajat will this also fix this deploy blocker?

Confirmed on iOS safari that it does.

roryabraham reviewed (changes requested):
Also, I must be going crazy because I can't find where the `isFocused` prop of the `ReportActionCompose` comes from, because it isn't coming from [here](https://github.com/Expensify/Expensify.cash/blob/37465dbd07da1feab8347835d82ed3d2302cde4c/src/pages/home/report/ReportView.js#L37). Are we sure that this is ever being actually set? Can we safely get rid of it? We also have `isFocused` in the `ReportActionCompose` state.

parasharrajat commented on src/pages/home/report/ReportActionContextMenu.js:
```diff
@@ -115,21 +116,24 @@ class ReportActionContextMenu extends React.Component {
                         this.props.reportID,
                         this.props.reportAction.reportActionID,
                         _.isEmpty(this.props.draftMessage) ? this.getActionText() : '',
                     );
```
Yes

parasharrajat commented:
`IsFocused` prop is coming from withNavigation HOC. It's the most important prop.  Removing it is not an option. :smile:

parasharrajat committed:
Merge branch 'main' of github.com:Expensify/Expensify.cash into edit-box

parasharrajat committed:
tweaks

parasharrajat committed:
typo

parasharrajat commented:
Updated. Thanks.

parasharrajat commented:
@deetergp @roryabraham Awaiting your feedback. Thanks.

deetergp commented:
## Testing
### Web
Editing via the context menu always puts focus to the edit text box, but the right-click > edit only does it on the first edit after a page load. In subsequent right-click > edits, the focus stays at the bottom, in the main text editor.

### Desktop
Context menu Edit works like it should, but I couldn't get right-click > edit to focus in the edit text box, it always went down to the main text box.

### mWeb
Works as intended 👍 

### iOS
Works as intended, though it's weird that you only get the long-press context menu if you click on your own avatar. That's probably unrelated.

parasharrajat committed:
fix: prop name

parasharrajat commented:
@deetergp Thanks for letting me know. I changed the prop name but forgot to update it in the Component. `shouldSetModalVisibility` was the prop. Just fixed it. You can check now. Thanks.

deetergp commented:
That fix did it, thanks @parasharrajat 👍

deetergp reviewed (approved):
Code looks good and tests pass.

parasharrajat commented:
:shipit:

parasharrajat commented:
@roryabraham Any feedback here? I think this PR is waiting to be merged by you. :sweat_smile:

roryabraham reviewed (approved):
This looks good to me. Sorry to have been a bottleneck here @parasharrajat – been a crazy couple days.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash/actions/runs/940035625) to staging in version: 1.0.68-5🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

kavimuru commented:
## iPhone - Safari -  Cursor focus not gained for edit comment field
## Expected Result: 
Textbox should automatically get focused.
## Actual Result:
Focus not gained in edit box
## Actions Performed:
1. Launch the URL and login
2. Hold press any text message sent by you and choose Edit Comment
## Platform:
iOS 
mWeb ✔️
Android 
## Build:
1.0.69-0

## Notes/Images/Video:

https://user-images.githubusercontent.com/43996225/122132148-50777b80-ce08-11eb-9be8-9bd8735cbc6f.mp4

parasharrajat commented:
No worries I will test it on safari M-web

roryabraham commented:
I am unable to reproduce on iOS 14.5 Safari. @isagoico @kavimuru can we please retest this in the latest staging version and see if it has been resolved?

parasharrajat commented:
Same here issue is not reproducible on M-web safari iPhone 12 - 14.4.

isagoico commented:
Retested and it was a pass in mWeb! Checking it off.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash/actions/runs/972294561) to production in version: 1.0.73-3🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
